tahoe_idp/models.py

```diff
@@ -0,0 +1,104 @@
+from urllib.parse import urlencode, urljoin
+
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import AbstractUser
+from django.db import models
+from django.http import HttpRequest
+from django.urls import reverse
+from django.utils import timezone
+from tahoe_idp import magiclink_settings
+from tahoe_idp.magiclink_utils import get_client_ip
+
+User = get_user_model()
+
+
+class MagicLinkError(Exception):
+    pass
+
+
+class MagicLink(models.Model):
+    username = models.CharField()
+    token = models.TextField()
+    expiry = models.DateTimeField()
+    redirect_url = models.TextField()
+    disabled = models.BooleanField(default=False)
+    times_used = models.IntegerField(default=0)
+    cookie_value = models.TextField(blank=True)
+    ip_address = models.GenericIPAddressField(null=True)
+    created = models.DateTimeField(auto_now_add=True)
+
+    def __str__(self):
+        return '{username} - {expiry}'.format(username=self.username, expiry=self.expiry)
+
+    def used(self) -> None:
+        self.times_used += 1
+        if self.times_used >= magiclink_settings.TOKEN_USES:
+            self.disabled = True
+        self.save()
+
+    def disable(self) -> None:
+        self.times_used += 1
+        self.disabled = True
+        self.save()
+
+    def generate_url(self, request: HttpRequest) -> str:
+        url_path = reverse(magiclink_settings.LOGIN_VERIFY_URL)
+
+        params = {'token': self.token}
+        if magiclink_settings.VERIFY_INCLUDE_USERNAME:
+            params['username'] = self.username
+        query = urlencode(params)
+
+        url_path = '{url_path}?{query}'.format(url_path=url_path, query=query)
+        scheme = request.is_secure() and 'https' or 'http'
+        url = urljoin(
+            '{scheme}://{studio_domain}'.format(scheme=scheme, studio_domain=magiclink_settings.STUDIO_DOMAIN),
+            url_path
+        )
+        return url
+
+    def validate(
+        self,
+        request: HttpRequest,
+        username: str = '',
+    ) -> AbstractUser:
+        if magiclink_settings.VERIFY_INCLUDE_USERNAME and self.username != username:
+            raise MagicLinkError('username does not match')
+
+        if timezone.now() > self.expiry:
```

Comment:
same `now` override to avoid flaky tests

```suggestion
        if now > self.expiry:
```

Reply:
patched in tests